torchviz/dot.py

```diff
@@ -6,68 +6,134 @@
-            if torch.is_tensor(var):
-                # note: this used to show .saved_tensors in pytorch0.2, but stopped
-                # working as it was moved to ATen and Variable-Tensor merged
-                dot.node(str(id(var)), size_to_str(var.size()), fillcolor='orange')
-            elif hasattr(var, 'variable'):
-                u = var.variable
-                name = param_map[id(u)] if params is not None else ''
-                node_name = '%s\n %s' % (name, size_to_str(u.size()))
-                dot.node(str(id(var)), node_name, fillcolor='lightblue')
-            elif var in output_nodes:
-                dot.node(str(id(var)), str(type(var).__name__), fillcolor='darkolivegreen1')
-            else:
-                dot.node(str(id(var)), str(type(var).__name__))
-            seen.add(var)
-            if hasattr(var, 'next_functions'):
-                for u in var.next_functions:
-                    if u[0] is not None:
-                        dot.edge(str(id(u[0])), str(id(var)))
-                        add_nodes(u[0])
-            if hasattr(var, 'saved_tensors'):
-                for t in var.saved_tensors:
-                    dot.edge(str(id(t)), str(id(var)))
-                    add_nodes(t)
+    def get_var_name(var, name=None):
+        name = param_map[id(var)] if id(var) in param_map else ''
+        return '%s\n %s' % (name, size_to_str(var.size()))
+
+    def add_nodes(fn):
+        if fn in seen:
+            return
+        seen.add(fn)
+
+        if show_saved:
+            for attr in dir(fn):
+                if not attr.startswith(SAVED_PREFIX):
+                    continue
+                val = getattr(fn, attr)
+                if torch.is_tensor(val):
+                    dot.edge(str(id(fn)), str(id(val)))
+                    dot.node(str(id(val)), attr[len(SAVED_PREFIX):], fillcolor='orange')
+                if isinstance(val, tuple):
+                    for i, t in enumerate(val):
+                        if torch.is_tensor(t):
+                            name = attr[len(SAVED_PREFIX):] + '[%s]' % str(i)
+                            dot.edge(str(id(fn)), str(id(t)))
+                            dot.node(str(id(t)), name, fillcolor='orange')
+
+        if hasattr(fn, 'variable'):
+            # if grad_accumulator, add the node for `.variable`
+            var = fn.variable
+            dot.node(str(id(var)), get_var_name(var), fillcolor='lightblue')
+            dot.edge(str(id(var)), str(id(fn)))
+
+        # add the node for this grad_fn
+        dot.node(str(id(fn)), get_fn_name(fn, show_attrs))
+
+        # recurse
+        if hasattr(fn, 'next_functions'):
+            for u in fn.next_functions:
+                if u[0] is not None:
+                    dot.edge(str(id(u[0])), str(id(fn)))
+                    add_nodes(u[0])
+
+        # note: this used to show .saved_tensors in pytorch0.2, but stopped
+        # working as it was moved to ATen and Variable-Tensor merged
```

Comment:
Can you add a comment saying that this still works for python custom autograd Function though.